pinot-common/src/main/java/org/apache/pinot/common/utils/DatabaseUtils.java

```
@@ -0,0 +1,91 @@
+import java.util.Objects;
+import javax.annotation.Nullable;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.core.MultivaluedMap;
+import org.apache.pinot.common.config.provider.TableCache;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class DatabaseUtils {
+  private DatabaseUtils() {
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatabaseUtils.class);
+
+  private static final List<String> TABLE_NAME_KEYS = List.of("tableName", "tableNameWithType", "schemaName");
+
+  public static void translateTableNameQueryParam(ContainerRequestContext requestContext, TableCache tableCache) {
+    MultivaluedMap<String, String> queryParams = requestContext.getUriInfo().getQueryParameters();
+    String uri = requestContext.getUriInfo().getRequestUri().toString();
+    String databaseName = null;
+    if (requestContext.getHeaders().containsKey(CommonConstants.DATABASE)) {
+      databaseName = requestContext.getHeaderString(CommonConstants.DATABASE);
+    }
+    for (String key : TABLE_NAME_KEYS) {
+      if (queryParams.containsKey(key)) {
+        String tableName = queryParams.getFirst(key);
+        String actualTableName = translateTableName(tableName, databaseName, tableCache);
+        // table is not part of default database
+        if (!actualTableName.equals(tableName)) {
+          uri = uri.replaceAll(String.format("%s=%s", key, tableName),
+              String.format("%s=%s", key, actualTableName));
+          try {
+            requestContext.setRequestUri(new URI(uri));
+          } catch (URISyntaxException e) {
+            LOGGER.error("Unable to translate the table name from {} to {}", tableName, actualTableName);
+          }
+        }
+      }
+    }
+  }
+
+  public static String translateTableName(String tableName, String databaseName, @Nullable TableCache tableCache) {
+    if (tableName != null && databaseName != null) {
```

This logic doesn't seem correct. When `database` is provided, why are we still trying to split the table name? What if the provided `database` doesn't match the table prefix?

---

database prefix in table name will take precedence over the database name passed explicitly. Will add java doc

---

Can we throw exception when they mismatch? Silently dropping one is dangerous.